### ISSUE TITLE
multiple metrics; confusion matrix?

### ISSUE THREAD

**Issue by stepthom** [microsoft/FLAML#160]
Are there any development plans to allow FLAML to report multiple metrics at the end of a `fit()` call? E.g., accuracy, f1, AUC, and even confusion matrix? My understanding is that you can only get `best_loss` which will correspond to the `metric` that was passed to `fit()`.

**Comment by sonichi** [comment]
There is no such plan so far. Do you have a suggestion?

**Comment by qingyun-wu** [comment]
Hi @stepthom, although we are not reporting multiple metrics at the end of a `fit()` call, there is actually a way to allow FLAML to log multiple metrics: you can define a [customized metric function ](https://github.com/microsoft/FLAML/blob/main/flaml/automl.py#L1020)which returns both a metric to minimize and a dictionary of metrics to log. Please find an example in this [test](https://github.com/microsoft/FLAML/blob/10082b9262a862ef51cf09117bc5d13e43d1ca0d/test/test_automl.py#L227). Once such a customized metric function is used, you can retrieve the multiple metrics of your interest in the log file.  And then reporting those metrics at the end of a `fit()` call should also be fairly easy. You are welcome to help to add it.  Thank you!

**Comment by stepthom** [comment]
@qingyun-wu Thank you! That looks very interesting. I will give it a try.

**Comment by stepthom** [comment]
@qingyun-wu @sonichi Thanks for your help thus far. I was able to write a custom metric that returns a lot of metrics (following the example in the tests):

```python
    def custom_metric(X_val, y_val, estimator, labels, X_train, y_train,
                  weight_val=None, weight_train=None):
        from sklearn.metrics import log_loss, accuracy_score, balanced_accuracy_score, f1_score, roc_auc_score
        import time
        
        start = time.time()
        y_pred_proba = estimator.predict_proba(X_val)
        y_pred = np.argmax(y_pred_proba,axis=1)
        y_pred = estimator.predict(X_val)
        pred_time = (time.time() - start) / len(X_val)
        
        custom_metrics = {
            'val_metric_log_loss':  log_loss(y_val, y_pred_proba, labels=labels, sample_weight=weight_val),
            'val_metric_accuracy':  accuracy_score(y_val, y_pred, sample_weight=weight_val),
            'val_metric_f1_micro':  f1_score(y_val, y_pred, sample_weight=weight_val, average="micro"),
            'val_metric_f1_macro':  f1_score(y_val, y_pred, sample_weight=weight_val, average="macro"),
            'val_metric_f1_weighted':  f1_score(y_val, y_pred, sample_weight=weight_val, average="weighted"),
            'val_metric_roc_auc':  roc_auc_score(y_val, y_pred_proba, sample_weight=weight_val, multi_class="ovo"),
        }
        
        y_pred_proba = estimator.predict_proba(X_train)
        y_pred = np.argmax(y_pred_proba,axis=1)
        y_pred = estimator.predict(X_train)
       
        custom_metrics.update({
            'train_metric_log_loss':  log_loss(y_train, y_pred_proba, labels=labels, sample_weight=weight_train),
            'train_metric_accuracy':  accuracy_score(y_train, y_pred, sample_weight=weight_train),
            'train_metric_f1_micro':  f1_score(y_train, y_pred, sample_weight=weight_train, average="micro"),
            'train_metric_f1_macro':  f1_score(y_train, y_pred, sample_weight=weight_train, average="macro"),
            'train_metric_f1_weighted':  f1_score(y_train, y_pred, sample_weight=weight_train, average="weighted"),
            'train_metric_roc_auc':  roc_auc_score(y_train, y_pred_proba, sample_weight=weight_train, multi_class="ovo"),
        })
        
        custom_metrics.update({
            "pred_time": pred_time,
        })
        
        val_loss = 1 - custom_metrics['val_metric_f1_micro']
        return val_loss, custom_metrics
```

Thanks again.
